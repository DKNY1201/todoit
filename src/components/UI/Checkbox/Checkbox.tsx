import * as React from 'react';

const Checkbox = () => (
    <div className="checkbox">
        a check box
    </div>
);

export default Checkbox;