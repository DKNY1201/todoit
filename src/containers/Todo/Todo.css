.todo-container {
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 10px;
}

.todo-container .todo-central {
    display: flex;
    justify-content: center;
    align-items: center;
}