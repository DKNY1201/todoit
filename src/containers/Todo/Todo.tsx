import * as React from 'react';

import Checkbox from '../../components/UI/Checkbox/Checkbox';
import Icon from '../../components/UI/Icon/Icon';
import './Todo.css';

export interface ITodoProp {
	completeTodo: () => void;
	content: string;
	schedule?: string;
}

interface ITodoState {
	isHovering: boolean;
}

class Todo extends React.Component<ITodoProp, ITodoState> {
	state = {
		isHovering: false
	}


	public render(): React.ReactNode {
		const {isHovering} = this.state;
		const {completeTodo, content, schedule} = this.props;

		return (
			<div className="todo-container" onMouseEnter={this.toggleIsHover} onMouseLeave={this.toggleIsHover}>
				{ isHovering && <Icon type="drag-drop" /> }
				<div className="todo-central">
					<Checkbox clicked={completeTodo} />
					<span className="todo-content">{content}</span>
					{ isHovering && <Icon type="comment" clicked={this.showCommentModal} /> }
					{
						!(!schedule && !isHovering) && <div className="todo-schedule">
							{ schedule ? schedule : isHovering ? <Icon type="schedule" clicked={this.showScheduleModal} /> : ''}
						</div>
					}


				</div>
				{ isHovering && <Icon type="action" clicked={this.showActionModal} /> }
			</div>
		);
	}

	private showCommentModal = () => {
		console.log('showCommentModal');
	}

	private showScheduleModal = () => {
		console.log('showScheduleModal');
	}

	private showActionModal = () => {
		console.log('showActionModal');
	}

	private toggleIsHover = () => {
		this.setState((prevState) => {
			return {
				isHovering:!prevState.isHovering
			}
		})
	}
}

// Todo.propTypes = {
// 	completeTodo: PropTypes.func.isRequired,
// 	content: PropTypes.string.isRequired,
// 	schedule: PropTypes.string.isRequired,
// };

export default Todo;