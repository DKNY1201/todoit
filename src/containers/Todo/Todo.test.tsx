import * as enzyme from 'enzyme';
import * as React from 'react';

import Checkbox from '../../components/UI/Checkbox/Checkbox';
import Todo from './Todo';

describe('<Todo /> rendering', () => {
	let wrapper;
	let wrapperInstance;

	beforeEach(() => {
		const content = "Default content";
		const isHovering = false;
		wrapper = enzyme.shallow(<Todo content={content} />);
		wrapper.setState({isHovering});
		wrapperInstance = wrapper.instance();
	});

	it('should render a checkbox component', () => {
		expect(wrapper.find(Checkbox).length).toEqual(1);
	});

	it('should render content correctly', () => {
		const content = "Bring clothes to Goodwill";
		wrapper.setProps({content});
		expect(wrapper.find('.todo-content').text()).toEqual(content);
	});

	it('shouldn\'t render Drag icon when not hover', () => {
		const isHovering = false;
		wrapper.setState({isHovering});
		expect(wrapper.find('.icon-drag-drop').length).toEqual(0);
	});

	it('should render Drag icon when hover', () => {
		const isHovering = true;
		wrapper.setState({isHovering});
		expect(wrapper.find('.icon-drag-drop').length).toEqual(1);
	});

	it('shouldn\'t render Action icon when not hover', () => {
		const isHovering = false;
		wrapper.setState({isHovering});
		expect(wrapper.find('.icon-action').length).toEqual(0);
	});

	it('should render Action icon when hover', () => {
		const isHovering = true;
		wrapper.setState({isHovering});
		expect(wrapper.find('.icon-action').length).toEqual(1);
	});

	it('shouldn\'t render Comment icon when not hover', () => {
		const isHovering = false;
		wrapper.setState({isHovering});
		expect(wrapper.find('.icon-comment').length).toEqual(0);
	});

	it('should render Comment icon when hover', () => {
		const isHovering = true;
		wrapper.setState({isHovering});
		expect(wrapper.find('.icon-comment').length).toEqual(1);
	});

	it('should render date/time? in Schedule area when schedule is set', () => {
		const schedule = '03/25/2019 22:42';
		wrapper.setProps({schedule});
		expect(wrapper.find('.todo-schedule').text()).toEqual(schedule);
	});

	it('shouldn\'t render Schedule area when schedule is not set and not hover', () => {
		const schedule = null;
		const isHovering = false;
		wrapper.setProps({schedule});
		wrapper.setState({isHovering});
		expect(wrapper.find('.todo-schedule').length).toEqual(0);
	});

	it('should render a Calendar in Schedule area when schedule is not set and is hovering', () => {
		const schedule = null;
		const isHovering = true;
		wrapper.setProps({schedule});
		wrapper.setState({isHovering});
		expect(wrapper.find('.todo-schedule').find('.icon-schedule').length).toEqual(1);
	});

	it('shouldn\'t be hovered at beginning', () => {
		expect(wrapperInstance.state.isHovering).toEqual(false);
	});
});

describe('<Todo /> events', () => {
	let wrapper, wrapperInstance;

	beforeEach(() => {
		const content = "Default content";
		const isHovering = false;
		wrapper = enzyme.shallow(<Todo content={content} />);
		wrapper.setState({isHovering});
		wrapperInstance = wrapper.instance();
	});

	it('should call toggleIsHover() and change isHover to true when mouse entered', () => {
		const spy = jest.spyOn(wrapperInstance, 'toggleIsHover');
		wrapperInstance.forceUpdate();
		expect(spy).toHaveBeenCalledTimes(0);
		wrapper.simulate('mouseover');
		expect(spy).toHaveBeenCalledTimes(1);
		expect(wrapperInstance.state.isHover).toEqual(true);
	});

	it('should call toggleIsHover() and change isHover to false when mouse leaved', () => {
		const spy = jest.spyOn(wrapperInstance, 'toggleIsHover');
		wrapperInstance.forceUpdate();
		expect(spy).toHaveBeenCalledTimes(0);
		wrapper.simulate('mouseleave');
		expect(spy).toHaveBeenCalledTimes(1);
		expect(wrapperInstance.state.isHover).toEqual(false);
	});

	// it('should call completeTodo() when check the checkbox', () => {
	// 	const spy = jest.spyOn(wrapperInstance, 'completeTodo');
	// 	wrapperInstance.forceUpdate();
	// 	expect(spy).toHaveBeenCalledTimes(0);
	// 	wrapper.find(Checkbox).simulate('click');
	// 	expect(spy).toHaveBeenCalledTimes(1);
	// });

	it('should call showCommentModal() when click on Comment icon', () => {
		const spy = jest.spyOn(wrapperInstance, 'showCommentModal');
		wrapperInstance.forceUpdate();
		expect(spy).toHaveBeenCalledTimes(0);
		wrapper.find('.todo-comment').simulate('click');
		expect(spy).toHaveBeenCalledTimes(1);
	});

	it('should call showScheduleModal() when click on Schedule icon', () => {
		const spy = jest.spyOn(wrapperInstance, 'showScheduleModal');
		wrapperInstance.forceUpdate();
		expect(spy).toHaveBeenCalledTimes(0);
		wrapper.find('.todo-schedule').simulate('click');
		expect(spy).toHaveBeenCalledTimes(1);
	});

	it('should call showActionModal() when click on Action icon', () => {
		const spy = jest.spyOn(wrapperInstance, 'showActionModal');
		wrapperInstance.forceUpdate();
		expect(spy).toHaveBeenCalledTimes(0);
		wrapper.find('.todo-action').simulate('click');
		expect(spy).toHaveBeenCalledTimes(1);
	});
});
