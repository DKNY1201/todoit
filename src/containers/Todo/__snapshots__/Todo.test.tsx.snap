// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo /> rendering should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Todo
    content="Default content"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        <div
          className="todo-central"
>
          <Checkbox
                    clicked={undefined}
          />
          <span
                    className="todo-content"
          >
                    Default content
          </span>
</div>,
        false,
      ],
      "className": "todo-container",
      "onMouseEnter": [Function],
      "onMouseLeave": [Function],
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Checkbox
              clicked={undefined}
/>,
            <span
              className="todo-content"
>
              Default content
</span>,
            false,
            false,
          ],
          "className": "todo-central",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "clicked": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Default content",
              "className": "todo-content",
            },
            "ref": null,
            "rendered": "Default content",
            "type": "span",
          },
          false,
          false,
        ],
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          <div
            className="todo-central"
>
            <Checkbox
                        clicked={undefined}
            />
            <span
                        className="todo-content"
            >
                        Default content
            </span>
</div>,
          false,
        ],
        "className": "todo-container",
        "onMouseEnter": [Function],
        "onMouseLeave": [Function],
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Checkbox
                clicked={undefined}
/>,
              <span
                className="todo-content"
>
                Default content
</span>,
              false,
              false,
            ],
            "className": "todo-central",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "clicked": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Default content",
                "className": "todo-content",
              },
              "ref": null,
              "rendered": "Default content",
              "type": "span",
            },
            false,
            false,
          ],
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
